internal/dialect/dialect.go

```diff
@@ -0,0 +1,180 @@
+package dialect
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/pressly/goose/v3/internal/dialect/dialectquery"
+)
+
+// Dialect is the type of database dialect.
+type Dialect string
+
+const (
+	Postgres   Dialect = "postgres"
+	Mysql      Dialect = "mysql"
+	Sqlite3    Dialect = "sqlite3"
+	Sqlserver  Dialect = "sqlserver"
+	Redshift   Dialect = "redshift"
+	Tidb       Dialect = "tidb"
+	Clickhouse Dialect = "clickhouse"
+	Vertica    Dialect = "vertica"
+)
+
+func NewStore(d Dialect, table string) (Store, error) {
+	if table == "" {
+		return nil, errors.New("table name cannot be empty")
+	}
+	var querier dialectquery.Querier
+	switch d {
+	case Postgres:
+		querier = dialectquery.NewPostgres(table)
+	case Mysql:
+		querier = dialectquery.NewMysql(table)
+	case Sqlite3:
+		querier = dialectquery.NewSqlite3(table)
+	case Sqlserver:
+		querier = dialectquery.NewSqlserver(table)
+	case Redshift:
+		querier = dialectquery.NewRedshift(table)
+	case Tidb:
+		querier = dialectquery.NewTidb(table)
+	case Clickhouse:
+		querier = dialectquery.NewClickhouse(table)
+	case Vertica:
+		querier = dialectquery.NewVertica(table)
+	default:
+		return nil, fmt.Errorf("unknown querier dialect: %v", d)
+	}
+	return &store{querier: querier}, nil
+}
+
+// Store is the interface that wraps the basic methods for a database dialect.
+//
+// A dialect is a set of SQL statements that are specific to a database.
+//
+// By defining a store interface, we can support multiple databases
+// with a single codebase.
+//
+// The underlying implementation does not modify the error returned by the
+// database driver. It is the callers responsibility to assert for the correct
+// error, such as sql.ErrNoRows.
+type Store interface {
```

Comment:
This also gets us one step closer to a fully context-aware package without having to instrument too many internal details.